contracts/zero-ex/contracts/src/features/OtcOrdersFeature.sol

```diff
@@ -0,0 +1,443 @@
+    /// @param takerTokenFillAmount Maximum taker token amount to
+    ///        fill this order with.
+    /// @return takerTokenFilledAmount How much maker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function _fillOtcOrderPrivate(
+        LibNativeOrder.OtcOrder memory order,
+        LibSignature.Signature memory makerSignature,
+        LibSignature.Signature memory takerSignature,
+        uint128 takerTokenFillAmount,
+        WethOptions wethOptions
+    )
+        private
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        LibNativeOrder.OtcOrderInfo memory orderInfo = getOtcOrderInfo(order);
+
+        // Must be fillable.
+        if (orderInfo.status != LibNativeOrder.OrderStatus.FILLABLE) {
+            LibNativeOrdersRichErrors.OrderNotFillableError(
+                orderInfo.orderHash,
+                uint8(orderInfo.status)
+            ).rrevert();
+        }
+
+        {
+            LibNativeOrdersStorage.Storage storage stor =
+                LibNativeOrdersStorage.getStorage();
+
+            // Must be fillable by the tx.origin.
+            if (
+                order.txOrigin != tx.origin &&
+                !stor.originRegistry[order.txOrigin][tx.origin]
+            ) {
+                LibNativeOrdersRichErrors.OrderNotFillableByOriginError(
+                    orderInfo.orderHash,
+                    tx.origin,
+                    order.txOrigin
+                ).rrevert();
+            }
+
+            // Maker signature must be valid for the order.
+            address makerSigner = LibSignature.getSignerOfHash(orderInfo.orderHash, makerSignature);
+            if (
+                makerSigner != order.maker &&
+                !stor.orderSignerRegistry[order.maker][makerSigner]
+            ) {
+                LibNativeOrdersRichErrors.OrderNotSignedByMakerError(
+                    orderInfo.orderHash,
+                    makerSigner,
+                    order.maker
+                ).rrevert();
+            }
+        }
+
+        address taker = msg.sender;
+        // If msg.sender is not the taker, validate the taker signature.
+        if (!_isSenderValidTaker(order.taker)) {
+            address takerSigner = LibSignature.getSignerOfHash(orderInfo.orderHash, takerSignature);
+            if (takerSigner != order.taker) {
+                LibNativeOrdersRichErrors.OrderNotSignedByTakerError(
+                    orderInfo.orderHash,
+                    takerSigner,
+                    order.taker
+                ).rrevert();
+            }
+            taker = order.taker;
+        }
+
+        // Settle between the maker and taker.
+        (takerTokenFilledAmount, makerTokenFilledAmount) = _settleOtcOrder(
+            order,
+            taker,
+            takerTokenFillAmount,
+            wethOptions
+        );
+
+        emit OtcOrderFilled(
```

Comment:
Could we avoid emitting events, and simply rely on the ERC20 transfers? would that save gas?

Reply:
cc @alexkroeger 👀

Reply:
but no, I think @sui414 would kill us